rs/execution_environment/src/execution_environment/tests/canister_task.rs

```diff
@@ -819,3 +821,337 @@ fn global_timer_produces_transient_error_on_out_of_cycles() {
+    )
+    .unwrap();
+
+    // Here we have:
+    // wasm_capacity = wasm_memory_limit = 20 Wasm Pages
+    // wasm_memory_threshold = 15 Wasm Pages
+
+    // Initially wasm_memory.size = 1
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(1)
+    );
+
+    test.ingress_raw(canister_id, "grow_mem", vec![]);
+    test.ingress_raw(canister_id, "grow_mem", vec![]);
+
+    // First ingress messages gets executed.
+    // wasm_memory.size = 1 + 7 = 8
+    // wasm_capacity - used_wasm_memory < self.wasm_memory_threshold
+    // Hook condition is triggered.
+    test.execute_slice(canister_id);
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(8)
+    );
+
+    // Though we have the second ingress message awaiting to be processed,
+    // hook will be executed first.
+    test.execute_slice(canister_id);
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(13)
+    );
+
+    // The second ingress message is executed after the hook.
+    test.execute_slice(canister_id);
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(20)
+    );
+}
+
+#[test]
+fn on_low_wasm_memory_is_executed_once() {
+    let mut test = ExecutionTestBuilder::new().build();
+
+    let update_grow_mem_size = 7;
+    let hook_grow_mem_size = 2;
+
+    let wat = get_wat_with_updade_and_hook_mem_grow(update_grow_mem_size, hook_grow_mem_size);
+
+    let canister_id = test.canister_from_wat(wat.as_str()).unwrap();
+
+    test.canister_update_wasm_memory_limit_and_wasm_memory_threshold(
+        canister_id,
+        (20 * WASM_PAGE_SIZE_IN_BYTES as u64).into(),
+        (15 * WASM_PAGE_SIZE_IN_BYTES as u64).into(),
+    )
+    .unwrap();
+
+    // Here we have:
+    // wasm_capacity = wasm_memory_limit = 20 Wasm Pages
+    // wasm_memory_threshold = 15 Wasm Pages
+
+    // Initially wasm_memory.size = 1
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(1)
+    );
+
+    // wasm_memory.size = 1 + 7 = 8
+    // wasm_capacity - used_wasm_memory < self.wasm_memory_threshold
+    // hence hook will be executed. After hook execution we have:
+    // wasm_memory.size = 8 + 2 = 10.
+    test.ingress(canister_id, "grow_mem", vec![]).unwrap();
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(10)
+    );
+
+    // wasm_memory.size = 10 + 7 = 17
+    // wasm_capacity - used_wasm_memory < self.wasm_memory_threshold
+    // but because the hook is already executed it will not be executed again.
+    test.ingress(canister_id, "grow_mem", vec![]).unwrap();
+
+    assert_eq!(
+        test.execution_state(canister_id).wasm_memory.size,
+        NumWasmPages::new(17)
+    );
+}
+
+#[test]
+fn on_low_wasm_memory_runs_after_dts_execution() {
```

Comment:
Not super clean in this test how we check that indeed DTS runs first before the hook. Maybe while a slice of DTS has been executed, we double check that memory has not grown beyond what's expected (so the hook hasn't run yet) and then only when the DTS finishes, the hook can only execute then.

Reply:
After DTS, we checked that the memory was exactly 1 + 7 (update method to grow it by 7). After we execute another slice which is the hook we check that wasm memory is 1 + 7 + 5 = 13. Since we grow memory only by 5 or 7 Pages, there is exactly 1 way to get a memory size of 8, and exactly one way to get a memory size of 13. Since we first check that memory is 8, and then 13, that ensures that the update is finished first and the hook is run afterward.